src/internal/streamstore/streamstore.go

```diff
@@ -99,3 +104,143 @@ func (di *streamItem) ToReader() io.ReadCloser {
 func (di *streamItem) Deleted() bool {
 	return false
 }
+
+// ---------------------------------------------------------------------------
+// common reader/writer/deleter
+// ---------------------------------------------------------------------------
+
+type backuper interface {
+	BackupCollections(
+		ctx context.Context,
+		bases []kopia.IncrementalBase,
+		cs []data.BackupCollection,
+		globalExcludeSet map[string]map[string]struct{},
+		tags map[string]string,
+		buildTreeWithBase bool,
+		errs *fault.Bus,
+	) (*kopia.BackupStats, *details.Builder, map[string]kopia.PrevRefs, error)
+}
+
+// write persists bytes to the store
+func write(
+	ctx context.Context,
+	tenantID string,
+	service path.ServiceType,
+	collectionPurpose string,
+	itemName string,
+	bup backuper,
+	mr Marshaller,
+	errs *fault.Bus,
+) (string, error) {
+	// construct the path of the container
+	p, err := path.Builder{}.
+		ToStreamStorePath(tenantID, collectionPurpose, service, false)
```

Comment:
tangential to this PR, but maybe eventually we should update the `BackupCollection` interface to return a `path.Builder` for `FullPath()` instead of a `path.Path`. That would technically make the requirements for this less strict so we wouldn't have to have a service in the path. Not sure if that would make it easy to generate "bad" paths in GC though

Reply:
Not against the idea, but I like the additional safety of the path.Path.  Spreading path.Builder around feels a little dangerous, considering the rigor we want out of pathing.  Also, nothing wrong with a little namespacing that includes the service.  Who knows when that might come in handy down the road?

Reply:
mmm I suppose so. I guess if we really feel a strong need for fewer prefix directories we can make another implementation of the `Path` interface that has less strict requirements? Everything in the kopia package and the details stuff should be independent of whether there's prefix stuff on the path